cpp/src/arrow/buffer.h

```diff
@@ -357,6 +371,15 @@ Status AllocateEmptyBitmap(MemoryPool* pool, int64_t length,
 ARROW_EXPORT
 Status AllocateEmptyBitmap(int64_t length, std::shared_ptr<Buffer>* out);
 
+/// \brief Allocate a Buffer which owns the memory held by a std::string
+///
+/// \param[in] data input data std::string
+/// \param[out] out the allocated buffer
+///
+/// \return Status message
+ARROW_EXPORT
+Status AllocateSTLStringBuffer(std::string& data, std::shared_ptr<STLStringBuffer>* out);
```

Comment:
Can you change the type of `data` to `const std::string&`?

Reply:
Good point. I did not make this as ``const``, since according to ``std::move`` semantics, the memory held by the original object will be in an `undefined` state after the move operation. So, essentially, we are mutating the state of the object post facto ``std::move``.

Reply:
Thanks, that makes sense.

Reply:
This isn't the desired scope of ARROW-3250. The idea was to be able to write `Buffer::FromString(std::move(str), &buffer)` and have the memory from `str` transferred to an `std::string` member of `buffer`